Multi directory libraries and executables

diml opened the pull request:
This PR adds the necessary support for multi-directory libraries and executables.

## Description

When writing `(include_subdirs true)` or simple `(include_subdirs)` in a `dune` file, dune will recurse in sub-directories to look for modules.

The recursion stops when encountering a sub-directory that:

- is part of another project
- contains `(include_subdirs true)` or `(include_subdirs)` 
- contains at least one module consumer stanza: `library`, `executable(s)`, `test(s)`

## Implementation

`Dir_contents` was adapted to recognize directories that are part of a group of directories. When `Dir_contents.get` is called on a directory that is part of a group, the whole group is scanned at once.

The rule producer in `gen_rules.ml` does exactly the same thing.

## TODO (before this can be merged)

- [x] when looking for .c/.h files, include all directories that are part of the group

## Future work

Add a mode `(include_subdirs qualified)``so that sub-directories are qualifed. i.e. file `a/b/x.ml` will be seen as `A.B.x` at the root, and as `B.x` in `a/`

jeremiedimino committed:
Implement scanning of multi-dir groups

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Wire things in gen_rules

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Implement the syntax

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Add tests

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
doc

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Make sure the include_subdirs stanza cannot appear twice

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

rgrinberg commented:
> Add a mode (include_subdirs qualified)``so that sub-directories are qualifed. i.e. filea/b/x.mlwill be seen asA.B.xat the root, and asB.xina/`

Shouldn't this be the default? IMO we shouldn't encourage people to create arbitrary grouping of the modules that isn't reflected in the module structure.

rgrinberg committed:
Move empty modules value to own definition

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented on src/dir_contents.ml:
```diff
@@ -434,23 +447,189 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+    | Standalone of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+    | Group_root of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
```
Adding a brief doc for every constructor here would be hlpeful

rgrinberg commented on src/dir_contents.ml:
```diff
@@ -434,23 +447,189 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+          Option.bind (Path.drop_build_context dir)
+            ~f:(File_tree.find_dir (Super_context.file_tree sctx))
+        with
+        | None -> Empty_standalone None
```
Is it intentional that this case doesn't populate the cache?

diml replied:
It does, line 520

rgrinberg commented on src/dir_contents.ml:
```diff
@@ -434,23 +447,189 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+
+let cache = Hashtbl.create 32
+
+let rec get sctx ~dir =
```
Why not use `Hashtbl.find_or_add` for this function (and others like it)

diml replied:
I guess we could use it here. It's just unfortunate that we end up systematically creating a closure while it's only useful in the not found case.

For other functions, we modify the cache in the not found case through a recursive call. It's not clear what the behavior of `Hashtbl.find_or_add` is in this case should be, so it seems better to avoid it.

jeremiedimino committed:
Document Dir_status.t

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

diml commented:
> Shouldn't this be the default? IMO we shouldn't encourage people to create arbitrary grouping of the modules that isn't reflected in the module structure.

True, although allowing unqualified subdirs is not much work and I believe will help porting older projects to dune. What about:

```
(include_subdirs <false|true|unqualified>)
```

?

diml commented:
> Btw, I cherry picked the refactoring commit and rebased the rest so that it applies cleanly on master. It should make things easier to review.

Thanks

rgrinberg commented:
> (include_subdirs <false|true|unqualified>)

What would `(include_subdirs false)` be useful for?

rgrinberg reviewed (approved):
I'm approving because the implementation looks good. @diml I'll let you pick what you think the default on qualified vs. unqualified should be.

diml commented:
Changing the default to `(include_subdirs true)` is a breaking change. For instance it breaks projects currently relying on `copy_files` to have multi-directory libraries. /cc @emillon as well in case he wants to chime in regarding the design (and/or code).

BTW, we still need to sort out the situation regarding .c/.h files before merging this. In the past there was a least one request to put .c files in a sub-directory. In the current state of this PR, .c/.h files in sub-directories are ignored, so we couldn't consider them in the future without introducing a breaking change. We can either make sure they are considered or simply forbid .c/.h files in sub-directories for now. I might do the latter for now since it's easier, in particular it's easier to setup the `-I` flags.

rgrinberg commented:
> Changing the default to (include_subdirs true) is a breaking change. For instance it breaks projects currently relying on copy_files to have multi-directory libraries.

Oh, I simply suggested that `(include_subdirs)` without an argument be equivalent to `(include_subdirs qualified)` rather than `(include_subdirs unqualified)`. Btw, now I'm thinking if it's best that we shouldn't use booleans at all as the argument here. It will be confusing as we are introducing a new feature that users aren't really used to.

diml commented:
That's a good point. I'll change the syntax to the following then:

```scheme
(include_subdirs <qualified|unqualified>)
```

diml commented:
And maybe we can leave out the short `(include_subdirs)` syntax for now, just to make things explicit.

jeremiedimino committed:
_

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Change the syntax as discussed

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Add a test for C stubs

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Support C files in sub-directories

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Make sure loading a group component loads the whole group

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

diml commented:
I changed the syntax and also implemented the support for C files in sub-directories, since it wasn't that hard. @rgrinberg I let you review the latest changes.

emillon commented on doc/dune-files.rst:
```diff
@@ -57,6 +57,11 @@ allowed to write an explicit ``Foo`` module, in which case this will
 
+Note that by default libraries and other things that consume
+OCaml/Reason modules only consume modules from the directory where the
+stanza appear. In order to delcare a multi-directory library, you need
```
typo: declare

emillon commented on doc/dune-files.rst:
```diff
@@ -787,6 +792,34 @@ A directory that is ignored will not be eagerly scanned by Dune. Any
+stop when encountering a sub-directory that:
+
+- is part of a different project (for instance when vendoring projects)
+- contains ``(include_subdirs true)`` or ``(include_subdirs)``
```
That should be `(include_subdirs unqualified)`

emillon replied:
& the short syntax has been removed if I understand correctly.

diml replied:
Yep. Given that it's a new thing, it seems better to make things explicit

emillon commented on src/dir_contents.ml:
```diff
@@ -179,12 +179,36 @@ type modules =
+  | Standalone -> [t]
+  | Group_root (lazy l) -> t :: l
+  | Group_part t ->
+    match t.kind with
```
The compiler seems to be doing fine, but I think adding `begin`/`end` around the nested `match` will make this easier to read. An alternative is to use deep pattern matching instead of the nested `match`, and move the `assert false` to the outer `match`.

diml replied:
I used a deep pattern matching

emillon commented on src/jbuild.ml:
```diff
@@ -1634,15 +1632,25 @@ module Documentation = struct
+
+  let t =
+    enum
+      [ "unqualified", Unqualified
```
One thing that seems a bit weird is that there is no syntax for the default case. Would it be useful to have `(include_subdirs no)`?

diml replied:
I suppose. It especially makes sense if we plan to one day change the default to `(include_subdirs qualified)`. I added `no` and reworked the documentation

jeremiedimino committed:
Fix doc

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
tweak

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

rgrinberg commented on src/dir_contents.ml:
```diff
@@ -426,23 +448,197 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+
+    | Is_component_of_a_group_but_not_the_root of
+        Super_context.Dir_with_jbuild.t option
+    (* Sub-directory of a directory with [(include_subdirs true)] *)
```
This commend needs to be updated as you've changed the argument for this stanza.

rgrinberg commented on src/dir_contents.ml:
```diff
@@ -426,23 +448,197 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+
+    | Group_root of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+    (* Directory with [(include_subdirs true)] *)
```
This comment as well

rgrinberg commented on src/gen_rules.ml:
```diff
@@ -238,13 +234,29 @@ module Gen(P : Install_rules.Params) = struct
+                Path.relative (Dir_contents.dir dc) fn :: acc
+              else
+                acc))
+        |> List.rev
```
Why do we need this reverse? I just see that we provide these as command line args, does the order matter there?

diml replied:
Indeed, it doesn't matter. Note that it's not even command line flags, it's just to specify dependencies

jeremiedimino committed:
Add a no mode

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

rgrinberg commented on src/gen_rules.ml:
```diff
@@ -704,11 +728,17 @@ module Gen(P : Install_rules.Params) = struct
+           | Some _ ->
+             match Dir_contents.kind (Dir_contents.get sctx ~dir) with
+             | Group_part root ->
+               SC.load_dir sctx ~dir:(Dir_contents.dir root)
```
How come we only do this in the case when there are no stanzas in `dir`? Seems like we should let the paren that is part of multi dir lib regardless

diml replied:
It was done in the other `gen_rules` function. However, it's true that this code wasn't very clear and was starting to become spaghetti code, so I refactored it

rgrinberg reviewed (changes requested):
Looks good. Just some comments that need to be updated first.

jeremiedimino committed:
Update comment

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Remove useless List.rev

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

jeremiedimino committed:
Refactor gen_rules

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

rgrinberg reviewed (approved):
Looks good. Just need a CHANGES update

jeremiedimino committed:
Update changelog

Signed-off-by: Jeremie Dimino <jeremie@dimino.org>

diml commented:
@bobot in the end we did add first class support for multi directory libraries

emillon commented:
Looks good, thanks!
